### PR TITLE
PHP7: use uopz for mocking in unit tests

### PR DESCRIPTION
**Pull request by macbre**
[PLATFORM-2138](https://wikia-inc.atlassian.net/browse/PLATFORM-2138)

This PR introduces changes that are compatible with both PHP 5.6 and 7.
- `WikiaMockProxyUopz` class introduced that is used by PHP7 - it uses `uopz` for mocking and modifying the behavior of the `new` operator. This class will be merged with `WikiaMockProxy` once we fully migrate to PHP7.
- `WikiaDispatcher` now calls `uopz_get_mock` to allow mocking of Nirvana controllers - `uopz` does not support classes mocking when `new` operator is provided with a class name by a string (see 00702ad)

@wladekb / @artursitarski 

Pinging @nandy-andy  and @saipetch to review 10588ab (`MonetizationModuleHelper` changes to make mocking possible in PHP7 - dynamic and static methods were mixed there)

### CODE REVIEW

File: extensions/wikia/ExactTargetUpdates/tests/client/ExactTargetClientTest.php [hunk 1]
```diff
@@ -202,6 +202,10 @@ public function testDefaultClientConstruction() {
 			return $getExactTargetClient->invoke( $client );
 		} );
 
+		# normalize resources to allow assertEquals to pass in PHP7
+		$expected->sdl = is_resource( $expected->sdl );
+		$soapClient->sdl = is_resource( $soapClient->sdl );
```
Comment (macbre) [inline extensions/wikia/ExactTargetUpdates/tests/client/ExactTargetClientTest.php:207]: Pinging @mixth-sense (touches code owned by @Wikia/sustaining-engineering-team)

File: extensions/wikia/CreateNewWiki/tests/CreateNewWikiControllerTest.php [hunk 2]
```diff
@@ -74,7 +74,7 @@ public function testCreateWikiSuccess( $testCase, $testData ) {
 			->will($this->returnValue($mainPageUrl));
 
 		$this->mockClass('CreateWiki', $createWiki);
-		$this->mockClass('GlobalTitle', $mainPageTitle);
+		$this->mockClass('GlobalTitle', $mainPageTitle, 'newFromText');
```
Comment (macbre) [inline extensions/wikia/CreateNewWiki/tests/CreateNewWikiControllerTest.php:82]: @Wikia/sustaining-engineering-team

File: extensions/wikia/CreateNewWiki/tasks/ConfigureWikiFactory.php [hunk 3]
```diff
@@ -82,7 +81,7 @@ public function getStaticVariables( $siteName, $imagesURL, $imagesDir, $dbName,
 			$wikiFactoryVariables['wgMetaNamespace'] = str_replace( [ ':', ' ' ], [ '', '_' ], $siteName );
 		}
 
-		wfGetLBFactory()->sectionsByDB[$dbName] = $wikiFactoryVariables['wgDBcluster'] = $wgCreateDatabaseActiveCluster;
+		wfGetLBFactory()->sectionsByDB[$dbName] = $wikiFactoryVariables['wgDBcluster'] = \F::app()->wg->CreateDatabaseActiveCluster;
```
Comment (macbre) [inline extensions/wikia/CreateNewWiki/tasks/ConfigureWikiFactory.php:84]: Pinging @Wikia/sustaining-engineering-team

File: extensions/wikia/CreateNewWiki/tests/tasks/ConfigureWikiFactoryTest.php [hunk 4]
```diff
@@ -17,6 +17,8 @@ public function setUp() {
 			'\Wikia\CreateNewWiki\Tasks\TaskContext', [ 'getLanguage', 'getWikiName' ]
 		);
 		$this->mockClass( 'TaskContext', $this->taskContextMock );
+
+		$this->mockGlobalVariable( 'wgCreateDatabaseActiveCluster', 'c7' );
```
Comment (macbre) [inline extensions/wikia/CreateNewWiki/tests/tasks/ConfigureWikiFactoryTest.php:21]: Pinging @Wikia/sustaining-engineering-team

**Comment by nandy-andy** [comment]
@macbre from my perspective the changes in https://github.com/Wikia/app/commit/10588abfa14699de9794eb26622428d515dbb85f look good Thanks :+1:

File: extensions/wikia/CreateNewWiki/tasks/ConfigureWikiFactory.php (hunk 3 above)
Comment (ghost) [inline extensions/wikia/CreateNewWiki/tasks/ConfigureWikiFactory.php:84]: 👍 LG, tej

File: extensions/wikia/CreateNewWiki/tests/CreateNewWikiControllerTest.php (hunk 2 above)
Comment (ghost) [inline extensions/wikia/CreateNewWiki/tests/CreateNewWikiControllerTest.php:82]: 👍 LG, tej

File: extensions/wikia/CreateNewWiki/tests/tasks/ConfigureWikiFactoryTest.php (hunk 4 above)
Comment (ghost) [inline extensions/wikia/CreateNewWiki/tests/tasks/ConfigureWikiFactoryTest.php:21]: 👍 LG, tej

File: extensions/wikia/ExactTargetUpdates/tests/client/ExactTargetClientTest.php (hunk 1 above)
Comment (ghost) [inline extensions/wikia/ExactTargetUpdates/tests/client/ExactTargetClientTest.php:207]: 👍 LG, tej - although this code is going to be retired completely in a matter of days.

File: extensions/wikia/WikiaMaps/tests/WikiaMapsParserTagControllerTest.php [hunk 5]
```diff
@@ -64,7 +64,7 @@ public function validateParseTagParamsDataProvider() {
 			[
 				'Invalid latitude parameter',
 				[ 'map-id' => 1, 'lat' => '0x32' ],
-				true
+				version_compare(PHP_VERSION, '7..0.', '>=') ? false : true # PHP7 does not cast strings to numbers
```
Comment (ghost) [inline extensions/wikia/WikiaMaps/tests/WikiaMapsParserTagControllerTest.php]: `7..0.` - is this correct?

File: includes/wikia/nirvana/WikiaDispatcher.class.php [hunk 6]
```diff
@@ -132,6 +132,15 @@ public function dispatch( WikiaApp $app, WikiaRequest $request ) {
 				$controller = new $controllerClassName; /* @var $controller WikiaController */
 				$response->setTemplateEngine($controllerClassName::DEFAULT_TEMPLATE_ENGINE);
 
+				// uopz can not override classes returned by new operator when the class name is passed as a string
+				global $wgRunningUnitTests;
```
Comment (ghost) [inline includes/wikia/nirvana/WikiaDispatcher.class.php:136]: This is hacky as hell, tej :-)

File: includes/wikia/tests/core/WikiaMockProxy.class.php [hunk 7]
```diff
@@ -278,7 +283,12 @@ public function disable() {
 
 		// disable this instance
 		$this->enabled = false;
-		unset_new_overload();
+
+		# TODO: remove when PHP7 migration is completed (PLATFORM-2138)
```
Comment (ghost) [inline includes/wikia/tests/core/WikiaMockProxy.class.php:287]: How will we find this when PHP7 migration is completed, tej?

Comment (macbre) [inline includes/wikia/tests/core/WikiaMockProxy.class.php:287]:
```
ack-grep "PHP7" --php | grep TODO
```

Seriously, there's a ticket for post-migration actions.

File: extensions/wikia/WikiaMaps/tests/WikiaMapsParserTagControllerTest.php (hunk 5 above)
Comment (macbre) [inline extensions/wikia/WikiaMaps/tests/WikiaMapsParserTagControllerTest.php]: Good catch, fixed in c4c4802

File: extensions/wikia/WikiaMaps/tests/WikiaMapsParserTagControllerTest.php [hunk 8]
```diff
@@ -64,7 +64,7 @@ public function validateParseTagParamsDataProvider() {
 			[
 				'Invalid latitude parameter',
 				[ 'map-id' => 1, 'lat' => '0x32' ],
-				true
+				version_compare(PHP_VERSION, '7.0.0.', '>=') ? false : true # PHP7 does not cast strings to numbers
```
Comment (wladekb) [inline extensions/wikia/WikiaMaps/tests/WikiaMapsParserTagControllerTest.php]: @macbre you meant only strings with hex numbers, didn't you?

Comment (macbre) [inline extensions/wikia/WikiaMaps/tests/WikiaMapsParserTagControllerTest.php]: Yep - 437510edc93cda7b3

File: includes/wikia/nirvana/WikiaDispatcher.class.php (hunk 6 above)
Comment (wladekb) [inline includes/wikia/nirvana/WikiaDispatcher.class.php:136]: https://github.com/krakjoe/uopz/issues/42 should help us avoid this hack

File: includes/wikia/services/tests/DiffEngineTest.php [hunk 9]
```diff
@@ -13,6 +13,6 @@ public function testGetDiff() {
 		$textB = 'foobar';
 
 		$diff = $engine->generateDiffBody( $textA, $textB );
-		$this->assertContains( '<span class="diffchange diffchange-inline">foo</span>', $diff );
+		$this->assertContains( 'class="diffchange diffchange-inline">foo</', $diff );
```
Comment (wladekb) [inline includes/wikia/services/tests/DiffEngineTest.php:16]: @macbre what's changed here?

Comment (macbre) [inline includes/wikia/services/tests/DiffEngineTest.php:16]: `<ins>` and `<del>` tags are used by `wikidiff2` extension version for PHP7

File: includes/wikia/tests/core/WikiaBaseTest.class.php [hunk 10]
```diff
@@ -92,7 +92,7 @@ protected function setUp() {
 		if ( $this->mockProxy !== null ) {
 			throw new Exception("Previous test did not execute tearDown()");
 		}
-		$this->mockProxy = new WikiaMockProxy();
+		$this->mockProxy = function_exists('runkit_method_rename') ? new WikiaMockProxy() : new WikiaMockProxyUopz();
```
Comment (wladekb) [inline includes/wikia/tests/core/WikiaBaseTest.class.php]: Imagine having php5.6 without runkit but with uopz enabled. I'd rather base this check on php version

File: includes/wikia/tests/core/WikiaMockProxy.class.php (hunk 7 above)
Comment (wladekb) [inline includes/wikia/tests/core/WikiaMockProxy.class.php:287]: Sure thing, dude!

**Comment by wladekb** [comment]
overall looks promising, I left a few comments inlined

File: includes/wikia/tests/core/WikiaBaseTest.class.php (hunk 10 above)
Comment (macbre) [inline includes/wikia/tests/core/WikiaBaseTest.class.php]: Used `version_compare` in 0e87368

**Pull request merged**